File: apps/desktop/src/lib/module-tree.ts
import { invoke as tauriInvoke } from '@tauri-apps/api/core';
import { CURRENT_PROTOCOL_VERSION, type InvokeCommand } from './health';

const STABLE_ID_PATTERN = /^[0-9a-f]{64}$/;
const COUNT_PATTERN = /^(?:0|[1-9][0-9]{0,19})$/;
const HEX_PATTERN = /^(?:[0-9a-f]{2})+$/;
const MAX_U64 = BigInt('18446744073709551615');
const MAX_PATH_BYTES = 131_072;
const MAX_ENTRIES = 100;
const MAX_DISPLAY_CHARACTERS = 256;

export interface ModuleTreeQueryV1 {
  afterModuleId: string | null;
  limit: number;
  parentModuleId: string | null;
}

export interface QueryModuleTreeRequestV1 extends ModuleTreeQueryV1 {
  protocolVersion: typeof CURRENT_PROTOCOL_VERSION;
}

export type ModuleTreeEntryKindV1 = 'manifestBoundary' | 'pathBoundary';
export type ModuleTreeChildStateV1 = 'leaf' | 'hasChildren';

export interface ModuleTreeRevisionV1 {
  contentHash: string;
  pathHex: string;
}

export interface ModuleTreeBoundaryEvidenceV1 {
  manifestRevision: ModuleTreeRevisionV1 | null;
  representativeRevision: ModuleTreeRevisionV1 | null;
}

export interface ModuleTreeFeatureCountV1 {
  count: string;
  truncated: boolean;
}

export interface ModuleTreeEntryV1 {
  boundaryEvidence: ModuleTreeBoundaryEvidenceV1;
  centralSymbols: ModuleTreeFeatureCountV1;
  childState: ModuleTreeChildStateV1;
  entrypoints: ModuleTreeFeatureCountV1;
  fileCount: string;
  kind: ModuleTreeEntryKindV1;
  manifestCount: string;
  moduleId: string;
  name: string;
  nameTruncated: boolean;
  rootPathHex: string | null;
  symbolCount: string;
  tests: ModuleTreeFeatureCountV1;
}

export interface ModuleTreePageV1 {
  entries: ModuleTreeEntryV1[];
  graphCommunityCount: string;
  indexRunId: string;
  nextAfterModuleId: string | null;
  parentModuleId: string | null;
  primaryModuleCount: string;
  snapshotId: string;
}

export type ModuleTreeResultV1 =
  | { status: 'noProject' }
  | { status: 'noPublishedIndex' }
  | { status: 'projectionUnavailable' }
  | { page: ModuleTreePageV1; status: 'available' };

export interface ModuleTreeResponseV1 {
  protocolVersion: typeof CURRENT_PROTOCOL_VERSION;
  result: ModuleTreeResultV1;
}

const invokeThroughTauri: InvokeCommand = (command, arguments_) =>
  tauriInvoke<unknown>(command, arguments_);

export async function queryModuleTree(
  query: ModuleTreeQueryV1 = {
    afterModuleId: null,
    limit: 50,
    parentModuleId: null,
  },
  invokeCommand: InvokeCommand = invokeThroughTauri,
): Promise<ModuleTreeResponseV1> {
  if (!isQuery(query)) {
    throw new Error('Module tree query does not match the V1 schema.');
  }
  const request: QueryModuleTreeRequestV1 = {
    ...query,
    protocolVersion: CURRENT_PROTOCOL_VERSION,
  };
  const payload = await invokeCommand('query_module_tree', { request });
  return parseModuleTreeResponseV1(payload);
}

export function parseModuleTreeResponseV1(payload: unknown): ModuleTreeResponseV1 {
  if (
    !isRecord(payload) ||
    !hasExactKeys(payload, ['protocolVersion', 'result']) ||
    payload.protocolVersion !== CURRENT_PROTOCOL_VERSION
  ) {
    throw new Error('Module tree response does not match the V1 schema.');
  }
  return { protocolVersion: payload.protocolVersion, result: parseResult(payload.result) };
}

function parseResult(value: unknown): ModuleTreeResultV1 {
  if (!isRecord(value) || typeof value.status !== 'string') {
    throw new Error('Module tree response contains an invalid result.');
  }
  if (value.status === 'noProject' && hasExactKeys(value, ['status'])) {
    return { status: 'noProject' };
  }
  if (value.status === 'noPublishedIndex' && hasExactKeys(value, ['status'])) {
    return { status: 'noPublishedIndex' };
  }
  if (value.status === 'projectionUnavailable' && hasExactKeys(value, ['status'])) {
    return { status: 'projectionUnavailable' };
  }
  if (value.status === 'available' && hasExactKeys(value, ['page', 'status'])) {
    return { page: parsePage(value.page), status: 'available' };
  }
  throw new Error('Module tree response contains an invalid result.');
}

function parsePage(value: unknown): ModuleTreePageV1 {
  if (
    !isRecord(value) ||
    !hasExactKeys(value, [
      'entries',
      'graphCommunityCount',
      'indexRunId',
      'nextAfterModuleId',
      'parentModuleId',
      'primaryModuleCount',
      'snapshotId',
    ]) ||
    !isStableId(value.indexRunId) ||
    !isStableId(value.snapshotId) ||
    !isOptionalStableId(value.parentModuleId) ||
    !isCount(value.primaryModuleCount) ||
    !isCount(value.graphCommunityCount) ||
    !Array.isArray(value.entries) ||
    value.entries.length > MAX_ENTRIES ||
    !isOptionalStableId(value.nextAfterModuleId)
  ) {
    throw new Error('Module tree response contains an invalid page.');
  }
  const entries = value.entries.map(parseEntry);
  if (entries.some((entry, index) => index > 0 && entries[index - 1]!.moduleId >= entry.moduleId)) {
    throw new Error('Module tree entries are duplicated or unordered.');
  }
  if (
    value.nextAfterModuleId !== null &&
    (entries.length === 0 || entries[entries.length - 1]!.moduleId !== value.nextAfterModuleId)
  ) {
    throw new Error('Module tree response contains an invalid next cursor.');
  }
  if (
    BigInt(value.primaryModuleCount) < BigInt(entries.length) ||
    entries.some((entry) => entry.moduleId === value.parentModuleId)
  ) {
    throw new Error('Module tree response contains contradictory page counts or hierarchy.');
  }
  return {
    entries,
    graphCommunityCount: value.graphCommunityCount,
    indexRunId: value.indexRunId,
    nextAfterModuleId: value.nextAfterModuleId,
    parentModuleId: value.parentModuleId,
    primaryModuleCount: value.primaryModuleCount,
    snapshotId: value.snapshotId,
  };
}

function parseEntry(value: unknown): ModuleTreeEntryV1 {
  if (
    !isRecord(value) ||
    !hasExactKeys(value, [
      'boundaryEvidence',
      'centralSymbols',
      'childState',
      'entrypoints',
      'fileCount',
      'kind',
      'manifestCount',
      'moduleId',
      'name',
      'nameTruncated',
      'rootPathHex',
      'symbolCount',
      'tests',
    ]) ||
    !isStableId(value.moduleId) ||
    !isEntryKind(value.kind) ||
    !isOptionalRepositoryPathHex(value.rootPathHex) ||
    typeof value.name !== 'string' ||
    value.name.length === 0 ||
    Array.from(value.name).length > MAX_DISPLAY_CHARACTERS ||
    containsControl(value.name) ||
    typeof value.nameTruncated !== 'boolean' ||
    (value.nameTruncated && Array.from(value.name).length !== MAX_DISPLAY_CHARACTERS) ||
    !isCount(value.manifestCount) ||
    !isCount(value.fileCount) ||
    !isCount(value.symbolCount) ||
    !isChildState(value.childState)
  ) {
    throw new Error('Module tree response contains an invalid entry.');
  }
  const boundaryEvidence = parseBoundaryEvidence(value.boundaryEvidence);
  const centralSymbols = parseFeatureCount(value.centralSymbols);
  const entrypoints = parseFeatureCount(value.entrypoints);
  const tests = parseFeatureCount(value.tests);
  const symbolCount = BigInt(value.symbolCount);
  const manifestShape =
    value.kind === 'manifestBoundary'
      ? BigInt(value.manifestCount) > BigInt(0) && boundaryEvidence.manifestRevision !== null
      : value.manifestCount === '0' && boundaryEvidence.manifestRevision === null;
  const representativeShape =
    (symbolCount === BigInt(0)) === (boundaryEvidence.representativeRevision === null);
  if (
    !manifestShape ||
    !representativeShape ||
    BigInt(value.fileCount) > symbolCount ||
    [centralSymbols, entrypoints, tests].some((feature) => BigInt(feature.count) > symbolCount)
  ) {
    throw new Error('Module tree response contains contradictory entry evidence.');
  }
  return {
    boundaryEvidence,
    centralSymbols,
    childState: value.childState,
    entrypoints,
    fileCount: value.fileCount,
    kind: value.kind,
    manifestCount: value.manifestCount,
    moduleId: value.moduleId,
    name: value.name,
    nameTruncated: value.nameTruncated,
    rootPathHex: value.rootPathHex,
    symbolCount: value.symbolCount,
    tests,
  };
}

function parseBoundaryEvidence(value: unknown): ModuleTreeBoundaryEvidenceV1 {
  if (!isRecord(value) || !hasExactKeys(value, ['manifestRevision', 'representativeRevision'])) {
    throw new Error('Module tree response contains invalid boundary evidence.');
  }
  return {
    manifestRevision: parseOptionalRevision(value.manifestRevision),
    representativeRevision: parseOptionalRevision(value.representativeRevision),
  };
}

function parseOptionalRevision(value: unknown): ModuleTreeRevisionV1 | null {
  if (value === null) return null;
  if (
    !isRecord(value) ||
    !hasExactKeys(value, ['contentHash', 'pathHex']) ||
    !isStableId(value.contentHash) ||
    !isRepositoryPathHex(value.pathHex)
  ) {
    throw new Error('Module tree response contains an invalid file revision.');
  }
  return { contentHash: value.contentHash, pathHex: value.pathHex };
}

function parseFeatureCount(value: unknown): ModuleTreeFeatureCountV1 {
  if (
    !isRecord(value) ||
    !hasExactKeys(value, ['count', 'truncated']) ||
    !isCount(value.count) ||
    typeof value.truncated !== 'boolean' ||
    (value.truncated && value.count === '0')
  ) {
    throw new Error('Module tree response contains an invalid feature count.');
  }
  return { count: value.count, truncated: value.truncated };
}

function isQuery(value: ModuleTreeQueryV1): boolean {
  return (
    hasExactKeys(value as unknown as Record<string, unknown>, [
      'afterModuleId',
      'limit',
      'parentModuleId',
    ]) &&
    isOptionalStableId(value.parentModuleId) &&
    isOptionalStableId(value.afterModuleId) &&
    Number.isInteger(value.limit) &&
    value.limit >= 1 &&
    value.limit <= MAX_ENTRIES
  );
}

function isOptionalRepositoryPathHex(value: unknown): value is string | null {
  return value === null || isRepositoryPathHex(value);
}

function isRepositoryPathHex(value: unknown): value is string {
  if (
    typeof value !== 'string' ||
    value.length > MAX_PATH_BYTES * 2 ||
    !HEX_PATTERN.test(value) ||
    containsHexByte(value, '00')
  )
    return false;
  const segments = splitHexPath(value);
  return (
    segments.length > 0 &&
    segments.every((segment) => segment.length > 0 && segment !== '2e' && segment !== '2e2e')
  );
}

function splitHexPath(value: string): string[] {
  const segments: string[] = [];
  let start = 0;
  for (let index = 0; index < value.length; index += 2) {
    if (value.slice(index, index + 2) === '2f') {
      segments.push(value.slice(start, index));
      start = index + 2;
    }
  }
  segments.push(value.slice(start));
  return segments;
}

function containsHexByte(value: string, target: string): boolean {
  for (let index = 0; index < value.length; index += 2) {
    if (value.slice(index, index + 2) === target) return true;
  }
  return false;
}

function isCount(value: unknown): value is string {
  return typeof value === 'string' && COUNT_PATTERN.test(value) && BigInt(value) <= MAX_U64;
}

function isStableId(value: unknown): value is string {
  return typeof value === 'string' && STABLE_ID_PATTERN.test(value);
}

function isOptionalStableId(value: unknown): value is string | null {
  return value === null || isStableId(value);
}

function isEntryKind(value: unknown): value is ModuleTreeEntryKindV1 {
  return value === 'manifestBoundary' || value === 'pathBoundary';
}

function isChildState(value: unknown): value is ModuleTreeChildStateV1 {
  return value === 'leaf' || value === 'hasChildren';
}

function containsControl(value: string): boolean {
  return Array.from(value).some((character) => {
    const codePoint = character.codePointAt(0);
    return (
      codePoint !== undefined && (codePoint <= 0x1f || (codePoint >= 0x7f && codePoint <= 0x9f))
    );
  });
}

function isRecord(value: unknown): value is Record<string, unknown> {
  return typeof value === 'object' && value !== null && !Array.isArray(value);
}

function hasExactKeys(value: Record<string, unknown>, expected: string[]): boolean {
  const actual = Object.keys(value).sort();
  const sortedExpected = [...expected].sort();
  return (
    actual.length === sortedExpected.length &&
    actual.every((key, index) => key === sortedExpected[index])
  );
}

File: crates/a3-protocol/src/module_tree.rs
use crate::ProtocolVersion;
use serde::{Deserialize, Serialize};

/// Strict bounded request for one page of the active project's deterministic module tree.
#[derive(Clone, Debug, Deserialize, Eq, PartialEq, Serialize)]
#[serde(deny_unknown_fields, rename_all = "camelCase")]
pub struct QueryModuleTreeRequestV1 {
    protocol_version: ProtocolVersion,
    parent_module_id: Option<String>,
    after_module_id: Option<String>,
    limit: u16,
}

impl QueryModuleTreeRequestV1 {
    /// Creates a request whose stable IDs and limit are validated by the Rust command boundary.
    #[must_use]
    pub const fn new(
        protocol_version: ProtocolVersion,
        parent_module_id: Option<String>,
        after_module_id: Option<String>,
        limit: u16,
    ) -> Self {
        Self {
            protocol_version,
            parent_module_id,
            after_module_id,
            limit,
        }
    }

    /// Creates the first top-level page with the product default size.
    #[must_use]
    pub const fn root() -> Self {
        Self::new(ProtocolVersion::CURRENT, None, None, 50)
    }

    /// Returns the requested protocol version.
    #[must_use]
    pub const fn protocol_version(&self) -> ProtocolVersion {
        self.protocol_version
    }

    /// Returns the selected parent module, or None for top level.
    #[must_use]
    pub fn parent_module_id(&self) -> Option<&str> {
        self.parent_module_id.as_deref()
    }

    /// Returns the exclusive stable module cursor.
    #[must_use]
    pub fn after_module_id(&self) -> Option<&str> {
        self.after_module_id.as_deref()
    }

    /// Returns the untrusted page-size value for application validation.
    #[must_use]
    pub const fn limit(&self) -> u16 {
        self.limit
    }
}

/// Versioned progressive module-tree result selected from the Core-owned active project.
#[derive(Clone, Debug, Deserialize, Eq, PartialEq, Serialize)]
#[serde(deny_unknown_fields, rename_all = "camelCase")]
pub struct ModuleTreeResponseV1 {
    protocol_version: ProtocolVersion,
    result: ModuleTreeResultV1,
}

impl ModuleTreeResponseV1 {
    /// Creates the response used before a project is active.
    #[must_use]
    pub const fn no_project() -> Self {
        Self {
            protocol_version: ProtocolVersion::CURRENT,
            result: ModuleTreeResultV1::NoProject,
        }
    }

    /// Creates the response used before the first atomic publication.
    #[must_use]
    pub const fn no_published_index() -> Self {
        Self {
            protocol_version: ProtocolVersion::CURRENT,
            result: ModuleTreeResultV1::NoPublishedIndex,
        }
    }

    /// Creates the response used when a historical publication predates module schema V8.
    #[must_use]
    pub const fn projection_unavailable() -> Self {
        Self {
            protocol_version: ProtocolVersion::CURRENT,
            result: ModuleTreeResultV1::ProjectionUnavailable,
        }
    }

    /// Creates an available page from application-validated bounded values.
    #[must_use]
    pub fn available(page: ModuleTreePageV1) -> Self {
        Self {
            protocol_version: ProtocolVersion::CURRENT,
            result: ModuleTreeResultV1::Available {
                page: Box::new(page),
            },
        }
    }

    /// Returns the mutually exclusive project/publication result.
    #[must_use]
    pub const fn result(&self) -> &ModuleTreeResultV1 {
        &self.result
    }
}

/// Whether an active project and current deterministic module projection exist.
#[derive(Clone, Debug, Deserialize, Eq, PartialEq, Serialize)]
#[serde(deny_unknown_fields, rename_all = "camelCase", tag = "status")]
pub enum ModuleTreeResultV1 {
    /// No project is active in this desktop process.
    NoProject,
    /// A project is active but no index has crossed the publish boundary.
    NoPublishedIndex,
    /// The latest historical publication has no deterministic module marker.
    ProjectionUnavailable,
    /// One bounded top-level or direct-child module page is available.
    Available {
        /// Current atomic page containing at most one hundred primary modules.
        page: Box<ModuleTreePageV1>,
    },
}

/// Bounded WebView-safe page of direct deterministic primary modules.
#[derive(Clone, Debug, Deserialize, Eq, PartialEq, Serialize)]
#[serde(deny_unknown_fields, rename_all = "camelCase")]
pub struct ModuleTreePageV1 {
    index_run_id: String,
    snapshot_id: String,
    parent_module_id: Option<String>,
    primary_module_count: String,
    graph_community_count: String,
    entries: Vec<ModuleTreeEntryV1>,
    next_after_module_id: Option<String>,
}

impl ModuleTreePageV1 {
    /// Creates one strict page from already validated application values.
    #[allow(clippy::too_many_arguments)]
    #[must_use]
    pub const fn new(
        index_run_id: String,
        snapshot_id: String,
        parent_module_id: Option<String>,
        primary_module_count: String,
        graph_community_count: String,
        entries: Vec<ModuleTreeEntryV1>,
        next_after_module_id: Option<String>,
    ) -> Self {
        Self {
            index_run_id,
            snapshot_id,
            parent_module_id,
            primary_module_count,
            graph_community_count,
            entries,
            next_after_module_id,
        }
    }
}

/// Deterministic signal that established a primary module boundary.
#[derive(Clone, Copy, Debug, Deserialize, Eq, PartialEq, Serialize)]
#[serde(rename_all = "camelCase")]
pub enum ModuleTreeEntryKindV1 {
    /// One or more current package manifests establish the boundary.
    ManifestBoundary,
    /// A deterministic repository path establishes the boundary.
    PathBoundary,
}

/// Whether another direct child-module page can exist below a node.
#[derive(Clone, Copy, Debug, Deserialize, Eq, PartialEq, Serialize)]
#[serde(rename_all = "camelCase")]
pub enum ModuleTreeChildStateV1 {
    /// No nested primary boundary exists.
    Leaf,
    /// At least one nested primary boundary exists.
    HasChildren,
}

/// Exact current file-revision evidence without source content.
#[derive(Clone, Debug, Deserialize, Eq, PartialEq, Serialize)]
#[serde(deny_unknown_fields, rename_all = "camelCase")]
pub struct ModuleTreeRevisionV1 {
    path_hex: String,
    content_hash: String,
}

impl ModuleTreeRevisionV1 {
    /// Creates one lossless relative path and content-hash pair.
    #[must_use]
    pub const fn new(path_hex: String, content_hash: String) -> Self {
        Self {
            path_hex,
            content_hash,
        }
    }
}

/// Representative membership and optional package-manifest evidence.
#[derive(Clone, Debug, Deserialize, Eq, PartialEq, Serialize)]
#[serde(deny_unknown_fields, rename_all = "camelCase")]
pub struct ModuleTreeBoundaryEvidenceV1 {
    representative_revision: Option<ModuleTreeRevisionV1>,
    manifest_revision: Option<ModuleTreeRevisionV1>,
}

impl ModuleTreeBoundaryEvidenceV1 {
    /// Creates evidence already validated against module kind and counts in Application.
    #[must_use]
    pub const fn new(
        representative_revision: Option<ModuleTreeRevisionV1>,
        manifest_revision: Option<ModuleTreeRevisionV1>,
    ) -> Self {
        Self {
            representative_revision,
            manifest_revision,
        }
    }
}

/// One bounded featured-symbol category and its visible truncation state.
#[derive(Clone, Debug, Deserialize, Eq, PartialEq, Serialize)]
#[serde(deny_unknown_fields, rename_all = "camelCase")]
pub struct ModuleTreeFeatureCountV1 {
    count: String,
    truncated: bool,
}

impl ModuleTreeFeatureCountV1 {
    /// Creates a lossless count with application-validated truncation semantics.
    #[must_use]
    pub const fn new(count: String, truncated: bool) -> Self {
        Self { count, truncated }
    }
}

/// One direct primary module with current deterministic evidence and bounded feature metadata.
#[derive(Clone, Debug, Deserialize, Eq, PartialEq, Serialize)]
#[serde(deny_unknown_fields, rename_all = "camelCase")]
pub struct ModuleTreeEntryV1 {
    module_id: String,
    kind: ModuleTreeEntryKindV1,
    root_path_hex: Option<String>,
    name: String,
    name_truncated: bool,
    boundary_evidence: ModuleTreeBoundaryEvidenceV1,
    manifest_count: String,
    file_count: String,
    symbol_count: String,
    central_symbols: ModuleTreeFeatureCountV1,
    entrypoints: ModuleTreeFeatureCountV1,
    tests: ModuleTreeFeatureCountV1,
    child_state: ModuleTreeChildStateV1,
}

impl ModuleTreeEntryV1 {
    /// Creates one application-validated primary module projection.
    #[allow(clippy::too_many_arguments)]
    #[must_use]
    pub const fn new(
        module_id: String,
        kind: ModuleTreeEntryKindV1,
        root_path_hex: Option<String>,
        name: String,
        name_truncated: bool,
        boundary_evidence: ModuleTreeBoundaryEvidenceV1,
        manifest_count: String,
        file_count: String,
        symbol_count: String,
        central_symbols: ModuleTreeFeatureCountV1,
        entrypoints: ModuleTreeFeatureCountV1,
        tests: ModuleTreeFeatureCountV1,
        child_state: ModuleTreeChildStateV1,
    ) -> Self {
        Self {
            module_id,
            kind,
            root_path_hex,
            name,
            name_truncated,
            boundary_evidence,
            manifest_count,
            file_count,
            symbol_count,
            central_symbols,
            entrypoints,
            tests,
            child_state,
        }
    }
}

#[cfg(test)]
mod tests {
    use super::{
        ModuleTreeBoundaryEvidenceV1, ModuleTreeChildStateV1, ModuleTreeEntryKindV1,
        ModuleTreeEntryV1, ModuleTreeFeatureCountV1, ModuleTreePageV1, ModuleTreeResponseV1,
        ModuleTreeRevisionV1, QueryModuleTreeRequestV1,
    };
    use crate::ProtocolVersion;

    #[test]
    fn available_page_serializes_counts_hierarchy_and_revision_evidence()
    -> Result<(), Box<dyn std::error::Error>> {
        let response = ModuleTreeResponseV1::available(ModuleTreePageV1::new(
            "11".repeat(32),
            "22".repeat(32),
            None,
            "1".to_owned(),
            "2".to_owned(),
            vec![ModuleTreeEntryV1::new(
                "33".repeat(32),
                ModuleTreeEntryKindV1::ManifestBoundary,
                None,
                "Repository".to_owned(),
                false,
                ModuleTreeBoundaryEvidenceV1::new(
                    Some(ModuleTreeRevisionV1::new(
                        "7372632f6c69622e7273".to_owned(),
                        "44".repeat(32),
                    )),
                    Some(ModuleTreeRevisionV1::new(
                        "436172676f2e746f6d6c".to_owned(),
                        "55".repeat(32),
                    )),
                ),
                "1".to_owned(),
                "1".to_owned(),
                "1".to_owned(),
                ModuleTreeFeatureCountV1::new("1".to_owned(), false),
                ModuleTreeFeatureCountV1::new("0".to_owned(), false),
                ModuleTreeFeatureCountV1::new("0".to_owned(), false),
                ModuleTreeChildStateV1::HasChildren,
            )],
            None,
        ));
        let value = serde_json::to_value(response)?;
        assert_eq!(value["result"]["status"], "available");
        assert_eq!(value["result"]["page"]["graphCommunityCount"], "2");
        assert_eq!(
            value["result"]["page"]["entries"][0]["boundaryEvidence"]["manifestRevision"]["contentHash"],
            "55".repeat(32)
        );
        assert_eq!(
            value["result"]["page"]["entries"][0]["childState"],
            "hasChildren"
        );
        Ok(())
    }

    #[test]
    fn request_rejects_unknown_fields_and_retains_untrusted_values_for_core_validation()
    -> Result<(), Box<dyn std::error::Error>> {
        let request = serde_json::from_value::<QueryModuleTreeRequestV1>(serde_json::json!({
            "protocolVersion": 1,
            "parentModuleId": "11",
            "afterModuleId": null,
            "limit": 100
        }))?;
        assert_eq!(request.protocol_version(), ProtocolVersion::V1);
        assert_eq!(request.parent_module_id(), Some("11"));
        assert_eq!(request.limit(), 100);

        let unknown = serde_json::json!({
            "protocolVersion": 1,
            "parentModuleId": null,
            "afterModuleId": null,
            "limit": 50,
            "repositoryPath": "C:/untrusted"
        });
        assert!(serde_json::from_value::<QueryModuleTreeRequestV1>(unknown).is_err());
        Ok(())
    }
}
